runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobInvocation.java

```diff
@@ -98,41 +99,71 @@ private FlinkJobInvocation(
+      PipelineResult runPipelineWithTranslator(FlinkPortablePipelineTranslator<T> translator)
+          throws Exception {
     LOG.info("Translating pipeline to Flink program.");
+
+    // Don't let the fuser fuse any subcomponents.
+    RunnerApi.Pipeline trimmedPipeline = makeKnownUrnsPrimitives(pipeline, translator.knownUrns());
+    System.out.println(trimmedPipeline.getComponents().getTransformsMap());
+
     // Fused pipeline proto.
-    RunnerApi.Pipeline fusedPipeline = GreedyPipelineFuser.fuse(pipeline).toPipeline();
+    RunnerApi.Pipeline fusedPipeline = GreedyPipelineFuser.fuse(trimmedPipeline).toPipeline();
     JobInfo jobInfo =
         JobInfo.create(
             id,
             pipelineOptions.getJobName(),
             retrievalToken,
             PipelineOptionsTranslation.toProto(pipelineOptions));
-    final JobExecutionResult result;
 
-    if (!pipelineOptions.isStreaming() && !hasUnboundedPCollections(fusedPipeline)) {
-      // TODO: Do we need to inspect for unbounded sources before fusing?
-      // batch translation
-      FlinkBatchPortablePipelineTranslator translator =
-          FlinkBatchPortablePipelineTranslator.createTranslator();
-      FlinkBatchPortablePipelineTranslator.BatchTranslationContext context =
-          FlinkBatchPortablePipelineTranslator.createTranslationContext(
-              jobInfo, pipelineOptions, confDir, filesToStage);
-      translator.translate(context, fusedPipeline);
-      result = context.getExecutionEnvironment().execute(pipelineOptions.getJobName());
-    } else {
-      // streaming translation
-      FlinkStreamingPortablePipelineTranslator translator =
-          new FlinkStreamingPortablePipelineTranslator();
-      FlinkStreamingPortablePipelineTranslator.StreamingTranslationContext context =
-          FlinkStreamingPortablePipelineTranslator.createTranslationContext(
-              jobInfo, pipelineOptions, confDir, filesToStage);
-      translator.translate(context, fusedPipeline);
-      result = context.getExecutionEnvironment().execute(pipelineOptions.getJobName());
-    }
+    FlinkPortablePipelineTranslator.Executor executor =
+        translator.translate(
+            translator.createTranslationContext(jobInfo, pipelineOptions, confDir, filesToStage),
+            fusedPipeline);
+    final JobExecutionResult result = executor.execute(pipelineOptions.getJobName());
 
     return FlinkRunner.createPipelineResult(result, pipelineOptions);
   }
 
+  private RunnerApi.Pipeline makeKnownUrnsPrimitives(
+      RunnerApi.Pipeline pipeline, Set<String> knownUrns) {
+    RunnerApi.Pipeline.Builder trimmedPipeline = pipeline.toBuilder();
+    for (String ptransformId : pipeline.getComponents().getTransformsMap().keySet()) {
+      if (knownUrns.contains(
+          pipeline.getComponents().getTransformsOrThrow(ptransformId).getSpec().getUrn())) {
+        System.out.println("FOUND " + ptransformId);
+        removeDescendants(trimmedPipeline, ptransformId);
```

Comment:
It would be nice if we didn't have to remove the descendants but simply recognize a transform as primitive during pipeline translation. I suppose this is not possible due to the Fuser merging stages. Makes me wonder, should we fix this in the Fuser instead?